File: app/src/main/java/com/gandhasiri/app/TreeDao.java
package com.gandhasiri.app;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.Query;
import androidx.room.Update;
import java.util.List;

@Dao
public interface TreeDao {
    @Insert
    long insert(Tree tree);

    @Update
    void update(Tree tree);

    @Delete
    void delete(Tree tree);

    @Query("SELECT * FROM trees ORDER BY datePlanted DESC")
    LiveData<List<Tree>> getAllTrees();

    @Query("SELECT * FROM trees WHERE id = :id")
    LiveData<Tree> getTreeById(int id);

    @Query("SELECT * FROM trees WHERE id = :id")
    Tree getTreeByIdSync(int id);

    @Insert
    void insertGrowthLog(GrowthLog log);

    @Query("SELECT * FROM growth_logs WHERE treeId = :treeId ORDER BY timestamp DESC")
    LiveData<List<GrowthLog>> getGrowthLogsForTree(int treeId);
}

File: app/src/main/java/com/gandhasiri/app/TreeDao.kt
package com.gandhasiri.app

import androidx.lifecycle.LiveData
import androidx.room.*

@Dao
interface TreeDao {
    @Insert
    fun insert(tree: Tree): Long

    @Update
    fun update(tree: Tree)

    @Delete
    fun delete(tree: Tree)

    @Query("SELECT * FROM trees")
    fun getAllTrees(): LiveData<List<Tree>>

    @Query("SELECT * FROM trees WHERE id = :id")
    fun getTreeById(id: Int): LiveData<Tree>

    @Query("SELECT * FROM trees WHERE id = :id")
    fun getTreeByIdSync(id: Int): Tree?

    @Insert
    fun insertGrowthLog(log: GrowthLog)

    @Query("SELECT * FROM growth_logs WHERE treeId = :treeId ORDER BY timestamp DESC")
    fun getGrowthLogsForTree(treeId: Int): LiveData<List<GrowthLog>>
}
